File: app/src/main/java/net/emojiparty/android/jishotomo/ui/csv/CsvExportAsyncTask.java
package net.emojiparty.android.jishotomo.ui.csv;

import android.content.Context;
import android.os.AsyncTask;
import net.emojiparty.android.jishotomo.data.csv.CsvExporter;
import net.emojiparty.android.jishotomo.ui.viewmodels.PagedEntriesControl;

public class CsvExportAsyncTask extends AsyncTask<Context, Integer, Void> {
  public interface CsvExportUiCallbacks {
    void onProgressUpdate(Integer progress);
    void onPreExecute();
    void onPostExecute();
    void onCanceled();
  }

  private CsvExportUiCallbacks uiCallbacks;
  private PagedEntriesControl pagedEntriesControl;

  public CsvExportAsyncTask(CsvExportUiCallbacks uiCallbacks, PagedEntriesControl pagedEntriesControl) {
    super();
    this.uiCallbacks = uiCallbacks;
    this.pagedEntriesControl = pagedEntriesControl;
  }

  @Override protected void onPreExecute() {
    super.onPreExecute();
    uiCallbacks.onPreExecute();
  }

  @Override protected void onPostExecute(Void aVoid) {
    super.onPostExecute(aVoid);
    uiCallbacks.onPostExecute();
  }

  @Override protected void onProgressUpdate(Integer... values) {
    super.onProgressUpdate(values);
    uiCallbacks.onProgressUpdate(values[0]);
  }

  @Override protected void onCancelled() {
    super.onCancelled();
    uiCallbacks.onCanceled();
  }

  @Override protected Void doInBackground(Context... contexts) {
    Context context = contexts[0];
    CsvExporter csvExporter = new CsvExporter(context);

    csvExporter.setCallback(new CsvExporter.ExportCallback() {
      @Override public void onUpdateProgress(Integer progress) {
        onProgressUpdate(progress);
      }

      @Override public void onFailure(Exception exception) {
        cancel(true);
      }
    });

    csvExporter.export(pagedEntriesControl.searchType,
        pagedEntriesControl.jlptLevel);

    return null;
  }
}

File: app/src/main/java/net/emojiparty/android/jishotomo/ui/csv/CsvExportAsyncTask.kt
package net.emojiparty.android.jishotomo.ui.csv

import android.content.Context
import android.os.AsyncTask
import net.emojiparty.android.jishotomo.data.csv.CsvExporter
import net.emojiparty.android.jishotomo.ui.viewmodels.PagedEntriesControl

class CsvExportAsyncTask(
  private val uiCallbacks: CsvExportUiCallbacks,
  private val pagedEntriesControl: PagedEntriesControl
) : AsyncTask<Context?, Int?, Void?>() {
  interface CsvExportUiCallbacks {
    fun onProgressUpdate(progress: Int?)
    fun onPreExecute()
    fun onPostExecute()
    fun onCancelled()
  }

  override fun onPreExecute() {
    super.onPreExecute()
    uiCallbacks.onPreExecute()
  }

  override fun onPostExecute(aVoid: Void?) {
    super.onPostExecute(aVoid)
    uiCallbacks.onPostExecute()
  }

  override fun onProgressUpdate(vararg values: Int?) {
    super.onProgressUpdate(*values)
    uiCallbacks.onProgressUpdate(values[0])

  }

  override fun onCancelled() {
    super.onCancelled()
    uiCallbacks.onCancelled()
  }

  override fun doInBackground(vararg params: Context?): Void? {
    params[0]?.let { context ->
      val csvExporter = CsvExporter(
          context,
          { progress ->
            onProgressUpdate(progress)
          },
          {
            cancel(true)
          }
      )

      csvExporter.export(pagedEntriesControl)
    }

    return null
  }
}
